src/AWS.Deploy.ServerMode.Client/ServerModeSession.cs

```diff
@@ -0,0 +1,243 @@
+using Amazon.Runtime;
+using Newtonsoft.Json;
+
+namespace AWS.Deploy.ServerMode.Client
+{
+    /// <summary>
+    /// Helper class that allows launching deployment tool in server mode.
+    /// It abstracts the server mode setup, CLI command execution and retries when desired port is unavailable.
+    /// </summary>
+    public interface IServerModeSession
+    {
+        /// <summary>
+        /// Starts deployment tool in server mode.
+        /// It creates symmetric key and tries to setup the server mode.
+        /// It also handles the retries when a desired port is unavailable in the provided range.
+        /// </summary>
+        /// <param name="cancellationToken">A cancellation token that can be used by other objects or threads to receive notice of cancellation.</param>
+        /// <exception cref="InternalServerModeException">Throws when deployment tool server failed to start for un unknown reason.</exception>
+        /// <exception cref="PortUnavailableException">Throws when deployment tool server failed to start due to unavailability of free ports.</exception>
+        Task Start(CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Builds <see cref="IRestAPIClient"/> client using the cached base URL.
+        /// If succeeded, <param name="restApiClient"></param> is initialized with current session client.
+        /// </summary>
+        /// <param name="credentialsGenerator">Func to that provides AWS credentials</param>
+        /// <param name="restApiClient"><see cref="IRestAPIClient"/> client to initialize.</param>
+        /// <returns>True, if <param name="restApiClient"/> is initialized successfully.</returns>
+        bool TryGetRestAPIClient(Func<Task<AWSCredentials>> credentialsGenerator, out IRestAPIClient? restApiClient);
+
+        /// <summary>
+        /// Builds <see cref="IDeploymentCommunicationClient"/> client using the cached base URL.
+        /// If succeeded, <param name="deploymentCommunicationClient"></param> is initialized with current session client.
+        /// </summary>
+        /// <param name="deploymentCommunicationClient"><see cref="DeploymentCommunicationClient"/> client to initialize.</param>
+        /// <returns>True, if <param name="deploymentCommunicationClient"/> is initialized successfully.</returns>
+        bool TryGetDeploymentCommunicationClient(out IDeploymentCommunicationClient? deploymentCommunicationClient);
+
+        /// <summary>
+        /// Returns the status of the deployment server by checking /api/v1/health API.
+        /// Returns true, if the deployment server returns a success HTTP code.
+        /// </summary>
+        /// <param name="cancellationToken">A cancellation token that can be used by other objects or threads to receive notice of cancellation.</param>
+        /// <returns>Returns true, if the deployment server returns a success HTTP code.</returns>
+        ValueTask<bool> IsAlive(CancellationToken cancellationToken);
```

Comment:
ValueTask will be hard for Visual Studio which is .NET Framework 4.6.1 to use. Switch to regular Task.

Reply:
Fixed.